### ISSUE TITLE
Test Project report is not proper when executing test case using data provider

### ISSUE THREAD

**Issue by VinothMessi** [testproject-io/java-sdk#62]
When am executing the test case one's we get a proper report with all entries

But when am executing the same test case multiple times using @DataProvider (different sets of test data), the report is not proper
[Distributed+Test+Executions-Job-Report-17-12-2020+08-39-28-full.pdf](https://github.com/testproject-io/java-sdk/files/5708102/Distributed%2BTest%2BExecutions-Job-Report-17-12-2020%2B08-39-28-full.pdf)

**Comment by DavidG8168** [comment]
Hi @VinothMessi,

We will take a look, can you meanwhile provide your test or a sample where this behavior is happening?

**Comment by DavidG8168** [comment]
Hello @VinothMessi ,

Can you please share your test? 

I could not reproduce with the following:

`

import io.testproject.sdk.drivers.web.ChromeDriver;
import org.openqa.selenium.chrome.ChromeOptions;
import org.testng.annotations.*;

public class InferredReportTest {

    /**
     * Driver instance.
     */
    private static ChromeDriver driver;
    private int urlNumber = 0;

    @BeforeTest
    static void setup() throws Exception {
        driver = new ChromeDriver("TOKEN",
                new ChromeOptions(), "dataProvider Annotation");
    }

    @DataProvider (name = "data-provider1")
    public Object[][] dpMethod1(){
        return new Object[][] {{"https://google.com"}, {"https://facebook.com"}, {"https://twitter.com"}};
    }

    @Test(testName = "Example Test 1", dataProvider = "data-provider1")
    void testExample1(String data) throws Exception {
        driver.navigate().to(data);
        driver.report().step("Navigated to URL #" + urlNumber);
        urlNumber++;
    }

    @DataProvider (name = "data-provider2")
    public Object[][] dpMethod2() {
        return new Object[][]{{"https://youtube.com"}, {"https://testproject.io"}};
    }

    @Test(testName = "Example Test 2", dataProvider = "data-provider2")
    void testExample2(String data) {
        driver.navigate().to(data);
        driver.report().step("Navigated to URL #" + urlNumber);
        urlNumber++;
    }

    @AfterTest
    static void tearDown() {
        if (driver != null) {
            driver.quit();
        }
    }
}

`
Here I essentially received a report containing two tests, since the @dataProvider annotation iteration doesn't count as a separate test execution, the steps were appended into the same test report.

If you'd like to receive a separate test for each of the iterations, you can create the driver and quit the driver as part of the @Test  method, like this:

`
 @Test(testName = "Example Test 1", dataProvider = "data-provider1")
    void testExample1(String data) throws Exception {
        driver = new ChromeDriver("TOKEN",
                new ChromeOptions(), "dataProvider Annotation");
        driver.navigate().to(data);
        driver.report().step("Navigated to URL #" + urlNumber);
        urlNumber++;
        driver.quit();
    }
`
Or opt to disabling auto-test reports, and deciding yourself when a test should be reported:
// After initializing the driver.
 driver.report().disableTestAutoReports(true);

// Whenever you decide a test should be submitted.
driver.report().test("End of Test", true, "Test has ended").submit();

In my test above it would be something like this:

`
@Test(testName = "Example Test 1", dataProvider = "data-provider1")
    void testExample1(String data) throws Exception {
        driver.navigate().to(data);
        driver.report().step("Navigated to URL #" + urlNumber);
        urlNumber++;
        driver.report().test("Example Test 1 Iteration" + urlNumber, true, "End of iteration").submit();
    }
`

**Comment by VinothMessi** [comment]
Hi David,

           This is my project, were i have used Data Provider and also have attached my test project report
Am executing the tests using open sdk: https://github.com/VinothMessi/java.web.automator.git

[Distributed+Test+Executions-Job-Report-24-12-2020+01-25-43-full.pdf](https://github.com/testproject-io/java-sdk/files/5739763/Distributed%2BTest%2BExecutions-Job-Report-24-12-2020%2B01-25-43-full.pdf)

**Comment by VinothMessi** [comment]
Any Update??? Please help me here...

**Comment by DavidG8168** [comment]
Hi @VinothMessi I am still checking this, 

Meanwhile, if you want your reports to look like this:

![image](https://user-images.githubusercontent.com/45967111/103214138-58099d80-4918-11eb-91f9-46cea4ff5364.png)

In the function:
`    public void flightReservation(Flightdetails fd) {
`
Make the following chances:
1. Before running your flow with ReservationBusinessFlow, disable test auto reports: `rDriver.disableTestAutoReports(true);`
2. After the flow, submit the report:` rDriver.test("MyTestName", booleanPassed).submit();`

Should be like this in total:
```
 @Test(testName = "Flight Reservation", description = "Flight Reservation Business Flow", dataProvider = "FlightReservation")
    public void flightReservation(Flightdetails fd) {
        rDriver.disableTestAutoReports(true);
        ReservationBusinessFlow.homepage(p -> {
    	// Code here.
        });
        rDriver.test("Test", true).submit();
    }
```

**Comment by VinothMessi** [comment]
{

    @Autowired
    private WebDriver browser;

    @Value("${app.url}")
    private String appUrl;

    @Object
    private FlightdetailsRepository flightdetailsRepository;

    @Object
    private FlightReservationBusinessFlow ReservationBusinessFlow;

    @Value("${app.url}")
    private String appURL;

    @Test(dataProvider = "FlightReservation")
    public void flightReservation(Flightdetails fd) {
        ((ReportingDriver) browser).report().disableTestAutoReports(true);
        //((ReportingDriver) browser).report().disableCommandReports(true);
        ReservationBusinessFlow.homepage(p -> {
            p.goTo(appURL);
            assertThat(p.hasItLoaded()).isTrue();
            ((ReportingDriver) browser).report().step("Navigate To Flight Application", "Navigated to Flight Application Successfully", true);
            System.out.println("Page Title: " + p.getPageHeader());
            ((ReportingDriver) browser).report().step("Page Header", "Grabbed Page Header Successfully", true);
            p.startRegistration();
        }).registration(p -> {
            assertThat(p.hasItLoaded()).isTrue();
            ((ReportingDriver) browser).report().step("Open Registration Page", "Registration Page Opened Successfully", true);
            System.out.println("Page Header: " + p.getPageHeader());
            System.out.println("Page Description: " + p.getPageDescription());
            p.getContactInformation().hasItLoaded();
            p.getContactInformation().fillContactInformations(fd.getFirstName(), fd.getLastName(), fd.getPhone(), fd.getEmail());
            ((ReportingDriver) browser).report().step("Contact Informations", "Contact Informations Filled Successfully", true);
            p.getMailingInformation().hasItLoaded();
            p.getMailingInformation().fillMailingInformations(fd.getAddress(), fd.getCity(), fd.getState(), fd.getPostalCode(),
                    fd.getCountry());
            ((ReportingDriver) browser).report().step("Mailing Informations", "Mailing Informations Filled Successfully", true);
            p.getUserInformation().hasItLoaded();
            p.getUserInformation().fillUserInformations(fd.getUserName(), fd.getPassword(), fd.getConfirmPassword());
            ((ReportingDriver) browser).report().step("User Informations", "User Informations Filled Successfully", true);
            p.goToRegistrationConfirmationPage();
        }).confirmation(p -> {
            assertThat(p.hasItLoaded()).isTrue();
            ((ReportingDriver) browser).report().step("Open Registration Confirmation Page", "Registration Confirmation Page Opened Successfully", true);
            System.out.println(p.getPageHeader());
            p.goToFlightDetailsPage();
        }).flightdetails(p -> {
            assertThat(p.hasItLoaded()).isTrue();
            ((ReportingDriver) browser).report().step("Open Flight Details Page", "Flight Details Page Opened Successfully", true);
            System.out.println("Page Header: " + p.getPageHeader());
            System.out.println("Page Description: " + p.getPageDescription());
            p.getFlightDetails().hasItLoaded();
            p.getFlightDetails().chooseTripType(fd.getTripType());
            p.getFlightDetails().selectNoOfPassengers(fd.getPassengers());
            p.getFlightDetails().departingDetails(fd.getDepartingPlace(), fd.getDepartingMonth(), fd.getArrivingDate());
            p.getFlightDetails().arrivingDetails(fd.getArrivingPlace(), fd.getArrivingMonth(), fd.getArrivingDate());
            p.getPreferences().selectAirline(fd.getAirline());
            ((ReportingDriver) browser).report().step("Flight Details Page", "Flight Details Filled Successfully", true);
            p.goToSelectFlightsPage();
        }).selectflights(p -> {
            assertThat(p.hasItLoaded()).isTrue();
            ((ReportingDriver) browser).report().step("Select Class", "Choosing Flight Service Class", true);
            p.goToBillingAddressPage();
        }).billingaddress(p -> {
            assertThat(p.hasItLoaded()).isTrue();
            p.fillInBillingAddress(fd.getAddress(), fd.getAddress(), fd.getCity(), fd.getState(), fd.getPostalCode());
            ((ReportingDriver) browser).report().step("Billing Address", "Filling in Billing Address Details", true);
            p.goToItineraryPage();
        }).flightitinerary(p -> {
            assertThat(p.hasItLoaded()).isTrue();
            ((ReportingDriver) browser).report().step("Flight Itinerary", "Flight Itinerary Confirmation", true);
            p.getConfirmationMessages();
            //p.shutDown();
        });
        ((ReportingDriver) browser).report().test("Flight Reservation Business Flow", true).submit();
    }

    @DataProvider(name = "FlightReservation")
    public java.lang.Object[] getContactInformations() {
        return flightdetailsRepository.findAll().stream().collect(Collectors.toList()).toArray();
    }

} 

It's working fine, but how can i use this in a different way? ((ReportingDriver) browser)

**Comment by DavidG8168** [comment]
What exactly are you referring to when you mention using it in a different way? 

The ReportingDriver as the namesake goes, is used for reporting.
Driver commands are already automatically reported, but you could use report.step() in methods you call more than once to avoid using it multiple times inside the test itself.

**Comment by VinothMessi** [comment]
If you see, i use WebDriver object from selenium interface. So for using the Test project reporting feature i typecast it as **((ReportingDriver) browser)**

Is there any alternative/better way other than this type casting?

**Comment by DavidG8168** [comment]
You'll need the cast if you you want to use the object from the Selenium interface.

I see you already have a Reporter object (rDriver) in ApplicationTests, you could use that to avoid the type cast each time.

**Comment by VinothMessi** [comment]
package com.web.application.e2e.automator.testcases;

import com.web.application.e2e.automator.ApplicationTests;
import com.web.application.e2e.automator.datajpa.entity.Flightdetails;

import io.testproject.sdk.drivers.ReportingDriver;
import org.testng.annotations.AfterMethod;
import org.testng.annotations.BeforeMethod;
import org.testng.annotations.DataProvider;
import org.testng.annotations.Test;

import java.util.stream.Collectors;

import static org.assertj.core.api.Assertions.*;

public class FlightReservation extends ApplicationTests {

    @Test(dataProvider = "FlightReservation")
    public void flightReservation(Flightdetails fd) {
        ReservationBusinessFlow.homepage(p -> {
            p.goTo(appUrl1);
            assertThat(p.hasItLoaded()).isTrue();
            ((ReportingDriver) browser).report().step("Navigate To Flight Application", "Navigated to Flight Application Successfully", true);
            System.out.println("Page Title: " + p.getPageHeader());
            ((ReportingDriver) browser).report().step("Page Header", "Grabbed Page Header Successfully", true);
            p.startRegistration();
        }).registration(p -> {
            assertThat(p.hasItLoaded()).isTrue();
            ((ReportingDriver) browser).report().step("Open Registration Page", "Registration Page Opened Successfully", true);
            System.out.println("Page Header: " + p.getPageHeader());
            System.out.println("Page Description: " + p.getPageDescription());
            p.getContactInformation().hasItLoaded();
            p.getContactInformation().fillContactInformations(fd.getFirstName(), fd.getLastName(), fd.getPhone(), fd.getEmail());
            ((ReportingDriver) browser).report().step("Contact Informations", "Contact Informations Filled Successfully", true);
            p.getMailingInformation().hasItLoaded();
            p.getMailingInformation().fillMailingInformations(fd.getAddress(), fd.getCity(), fd.getState(), fd.getPostalCode(),
                    fd.getCountry());
            ((ReportingDriver) browser).report().step("Mailing Informations", "Mailing Informations Filled Successfully", true);
            p.getUserInformation().hasItLoaded();
            p.getUserInformation().fillUserInformations(fd.getUserName(), fd.getPassword(), fd.getConfirmPassword());
            ((ReportingDriver) browser).report().step("User Informations", "User Informations Filled Successfully", true);
            p.goToRegistrationConfirmationPage();
        }).confirmation(p -> {
            assertThat(p.hasItLoaded()).isTrue();
            ((ReportingDriver) browser).report().step("Open Registration Confirmation Page", "Registration Confirmation Page Opened Successfully", true);
            System.out.println(p.getPageHeader());
            p.goToFlightDetailsPage();
        }).flightdetails(p -> {
            assertThat(p.hasItLoaded()).isTrue();
            ((ReportingDriver) browser).report().step("Open Flight Details Page", "Flight Details Page Opened Successfully", true);
            System.out.println("Page Header: " + p.getPageHeader());
            System.out.println("Page Description: " + p.getPageDescription());
            p.getFlightDetails().hasItLoaded();
            p.getFlightDetails().chooseTripType(fd.getTripType());
            p.getFlightDetails().selectNoOfPassengers(fd.getPassengers());
            p.getFlightDetails().departingDetails(fd.getDepartingPlace(), fd.getDepartingMonth(), fd.getArrivingDate());
            p.getFlightDetails().arrivingDetails(fd.getArrivingPlace(), fd.getArrivingMonth(), fd.getArrivingDate());
            p.getPreferences().selectAirline(fd.getAirline());
            ((ReportingDriver) browser).report().step("Flight Details Page", "Flight Details Filled Successfully", true);
            p.goToSelectFlightsPage();
        }).selectflights(p -> {
            assertThat(p.hasItLoaded()).isTrue();
            ((ReportingDriver) browser).report().step("Select Class", "Choosing Flight Service Class", true);
            p.goToBillingAddressPage();
        }).billingaddress(p -> {
            assertThat(p.hasItLoaded()).isTrue();
            p.fillInBillingAddress(fd.getAddress(), fd.getAddress(), fd.getCity(), fd.getState(), fd.getPostalCode());
            ((ReportingDriver) browser).report().step("Billing Address", "Filling in Billing Address Details", true);
            p.goToItineraryPage();
        }).flightitinerary(p -> {
            assertThat(p.hasItLoaded()).isTrue();
            ((ReportingDriver) browser).report().step("Flight Itinerary", "Flight Itinerary Confirmation", true);
            p.getConfirmationMessages();
            //p.shutDown();
        });

    }

    @DataProvider(name = "FlightReservation")
    public java.lang.Object[] getContactInformations() {
        return flightdetailsRepository.findAll().stream().collect(Collectors.toList()).toArray();
    }

    @BeforeMethod
    public void setUp() {
        ((ReportingDriver) browser).report().disableTestAutoReports(true);
        //((ReportingDriver) browser).report().disableCommandReports(true);
    }

    @AfterMethod
    public void tearDown() {
        ((ReportingDriver) browser).report().test("Flight Reservation Business Flow", true).submit();
    }

}

Look at the attached repot... Used the ((ReportingDriver) browser).report() in After Mthod, last test case entry alone is incorrect

[Distributed+Test+Executions-Job-Report-29-12-2020+12-46-02-full.pdf](https://github.com/testproject-io/java-sdk/files/5750725/Distributed%2BTest%2BExecutions-Job-Report-29-12-2020%2B12-46-02-full.pdf)

**Comment by DavidG8168** [comment]
You're right, add a short pause in the teardown:

    @AfterMethod
    public void tearDown() throws Exception {
        Thread.sleep(500);
        ((ReportingDriver) browser).report().test("Flight Reservation Business Flow", true).submit();
    }

The report should then appear as expected:

![image](https://user-images.githubusercontent.com/45967111/103288733-3da1f380-49ee-11eb-8ade-c2993f171995.png)

**Comment by VinothMessi** [comment]
Please do fix this issue as soon as possible

**Comment by DavidG8168** [comment]
#66

**Comment by DavidG8168** [comment]
This issue has been resolved on version 0.65.1.
